### PR TITLE
251 add test set for divi data

### LINKED ISSUE #251: Add test set for DIVI data
In the test for getDIVIData.py the data from DIVI is downloaded thrice but testing the functionality of divi should be done with a test data set.

### PR DESCRIPTION
**Pull request by annette-lutz**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [ ] Coverage report for new code is acceptable. 

Closes #251.

### CODE REVIEW

**Review by patricklnz — CHANGES_REQUESTED** [review]
Full review

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py [hunk 1]
```diff
@@ -74,14 +77,12 @@ def gdd_calls(self, text=''):
                 os.path.join(directory, 'germany_divi'+text+'.json'))]
         return gdd_calls
 
-    def test_extract_subframe_based_on_dates(self):
-        (df_raw, df_counties, df_states, df_ger) = gdd.get_divi_data(
-            out_folder=self.path)
+    def test_extracted_state_subframe(self):
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py:80]: I remove this test.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py [hunk 2]
```diff
@@ -124,22 +128,120 @@ def test_det_divi_data_prints(self, mock_print):
...
+        mock_san.assert_has_calls([call(self.df_test)])
+
+    @patch('memilio.epidata.getDIVIData.divi_data_sanity_checks')
+    @patch('memilio.epidata.getDIVIData.gd.loadCsv')
+    @patch('builtins.print')
+    def test_get_divi_data(self, mock_print, mock_csv, mock_san):
+        mock_csv.return_value = self.df_test
+        # test case with standard parameters
+        (df, df_county, df_states, df_ger) = gdd.get_divi_data(out_folder=self.path)
+        mock_san.assert_has_calls([call(self.df_test)])
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py:140]: This tests whether the `divi_data_sanity_checks` function which is mocked by `mock_san` is called with `self.df_test` as input.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py [hunk 3]
```diff
@@ -124,22 +128,120 @@ def test_det_divi_data_prints(self, mock_print):
...
-    def test_gdd_ma(self, mock_print):
+    def test_gdd_ma(self, mock_print, mock_csv, mock_san):
+        mock_csv.return_value = self.df_test
         # test case with moving average
-        gdd.get_divi_data(out_folder=self.path, moving_average=3)
-        expected_calls = self.gdd_calls('_ma3')
-        mock_print.assert_has_calls(expected_calls)
+        (df, df_county, df_states, df_ger) = gdd.get_divi_data(
+            out_folder=self.path, moving_average=3)
+        mock_san.assert_has_calls([call(self.df_test)])
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py:177]: see above

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py [hunk 4]
```diff
@@ -124,22 +128,120 @@ def test_det_divi_data_prints(self, mock_print):
         gdd_calls = self.gdd_calls(text='')
         expected_calls = expected_call + gdd_calls
         mock_print.assert_has_calls(expected_calls)
+        mock_san.assert_has_calls([call(self.df_test)])
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py:131]: See answer below.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py [hunk 5]
```diff
@@ -124,22 +128,120 @@ def test_det_divi_data_prints(self, mock_print):
...
-    def test_gdd_all_dates(self, mock_print):
+    def test_gdd_all_dates(self, mock_print, mock_csv, mock_san):
+        mock_csv.return_value = self.df_test
         # test case with impute dates is True
-        gdd.get_divi_data(out_folder=self.path, impute_dates=True)
-        expected_calls = self.gdd_calls(text='_all_dates')
-        mock_print.assert_has_calls(expected_calls)
+        (df, df_county, df_states, df_ger) = gdd.get_divi_data(
+            out_folder=self.path, impute_dates=True)
+        mock_san.assert_has_calls([call(self.df_test)])
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_get_divi_data.py:214]: see above

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_modifyDataframeSeries.py [hunk 6]
```diff
@@ -169,5 +192,49 @@ def test_impute_and_reduce_df(self):
...
 
 
+    def test_extract_subframe_based_on_dates(self):
+        test_df = self.df_dates.copy()
+        # test with start and end value in dataframe
+        extracted_df = mDfS.extract_subframe_based_on_dates(
+            self.df_dates, date(2022, 1, 3), date(2022, 1, 4))
+        pd.testing.assert_frame_equal(extracted_df, self.df_dates_result)
+        # check that input frame is unchanged
+        pd.testing.assert_frame_equal(self.df_dates, test_df)
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_modifyDataframeSeries.py:202]: If the dataframe is not copied but directly manipulated in the function `extract_subframe_based_on_dates`, this test fails.

File: pycode/memilio-epidata/memilio/epidata_test/test_epidata_modifyDataframeSeries.py [hunk 7]
```diff
@@ -169,5 +192,49 @@ def test_impute_and_reduce_df(self):
...
+        # test with start and end value not in dataframe
+        extracted_df = mDfS.extract_subframe_based_on_dates(
+            self.df_dates, date(2022, 1, 2), date(2022, 1, 5))
+        pd.testing.assert_frame_equal(extracted_df, self.df_dates_result)
+        pd.testing.assert_frame_equal(self.df_dates, test_df)
```
[the remark this answers is not in the document]
Comment (annette-lutz) [reply pycode/memilio-epidata/memilio/epidata_test/test_epidata_modifyDataframeSeries.py:207]: same as above

**Review by patricklnz — APPROVED** [review]
Looks good.

**Pull request merged**